jetbrains-core/src/software/aws/toolkits/jetbrains/services/lambda/python/PythonLambdaHandlerResolver.kt

```diff
@@ -3,46 +3,121 @@
+import com.intellij.psi.PsiDirectory
 import com.intellij.psi.PsiElement
 import com.intellij.psi.search.GlobalSearchScope
-import com.intellij.psi.util.QualifiedName
+import com.intellij.util.text.nullize
 import com.jetbrains.python.PyTokenTypes
 import com.jetbrains.python.psi.PyFile
 import com.jetbrains.python.psi.PyFunction
-import com.jetbrains.python.psi.PyPsiFacade
-import com.jetbrains.python.psi.resolve.fromModule
+import com.jetbrains.python.psi.stubs.PyModuleNameIndex
 import software.aws.toolkits.jetbrains.services.lambda.LambdaHandlerResolver
 
 class PythonLambdaHandlerResolver : LambdaHandlerResolver {
     override fun version(): Int = 1
 
-    override fun determineHandlers(element: PsiElement, file: VirtualFile): Set<String> = determineHandler(element)?.let { setOf(it) }.orEmpty()
+    override fun determineHandlers(element: PsiElement, file: VirtualFile): Set<String> =
+        determineHandler(element)?.let { setOf(it) }.orEmpty()
 
     override fun findPsiElements(
         project: Project,
         handler: String,
         searchScope: GlobalSearchScope
     ): Array<NavigatablePsiElement> {
-        val psiFacade = PyPsiFacade.getInstance(project)
-        val lambdaModule = handler.substringBeforeLast(".")
-        val function = handler.substringAfterLast(".")
-        return ModuleManager.getInstance(project).modules.flatMap { module ->
-            psiFacade.resolveQualifiedName(
-                QualifiedName.fromDottedString(lambdaModule),
-                fromModule(module)
-            )
-                .filterIsInstance<PyFile>()
-                .flatMap { pyFile ->
-                    pyFile.children.filterIsInstance<NavigatablePsiElement>()
-                        .filter { psiFunction -> psiFunction.name == function }
-                }
-        }.toTypedArray()
+        // <f/old/ers>.<par.ent.module.s.<module>.<function>
+        val functionName = handler.substringAfterLast('.')
+        var remainder = handler.substringBeforeLast('.')
+
+        val moduleStartIndex = remainder.lastIndexOfAny(charArrayOf('.', '/'))
+        if (moduleStartIndex < 0) return NavigatablePsiElement.EMPTY_NAVIGATABLE_ELEMENT_ARRAY
+
+        val moduleName = remainder.substring(moduleStartIndex + 1)
+        val isReferencedByModule = remainder[moduleStartIndex] == '.'
+
+        remainder = remainder.substring(0, remainder.length - moduleName.length - 1)
+
+        val parentModules = if (isReferencedByModule) {
+            remainder.substringAfterLast("/").split('.')
+        } else {
+            emptyList()
+        }
+
+        val parentFolders = if (isReferencedByModule) {
+            remainder.substringBeforeLast("/", "").nullize(true)?.split('/') ?: emptyList()
+        } else {
+            remainder.split('/')
+        }
+
+        // Find the module by the name
+        val modules = PyModuleNameIndex.find(moduleName, project, true)
+        for (pyModule in modules) {
```

Comment:
Nit: `modules.foreach` (unless the `continue` / `return` makes that unwieldy)

Reply:
I like the old school loop more than `return@reallyLongName` :-/

Reply:
Switched back to `.foreach`, originally the helper method was embedded in it so it had nested loops. Now that it is 1 layer deep, `.foreach` is clean again